### ISSUE TITLE
Boundary conditions in cubic splines

### ISSUE THREAD

**Issue by craabreu** [openmm/openmm#2627]
There are at least two issues concerning the boundary conditions used in the `SplineFitter` class for one-dimensional splines:

1. The method `createNaturalSpline` actually creates a clamped spline [[1](https://docs.scipy.org/doc/scipy-0.18.1/reference/generated/scipy.interpolate.CubicSpline.html#scipy.interpolate.CubicSpline), [2](https://en.wikipedia.org/wiki/Spline_interpolation)], whose first-order derivative is set to be zero at the endpoints. For a natural spline, this should be done for the second-order derivative instead.

1. The method `createPeriodicSpline` fails to produce a continuous first-order derivative at the boundary, which has implications for torsion forces with tabulated functions, such as `CMAPTorsionForce`. This can be attested by appending the following code lines to the `testPeriodicSpline` routine in `tests/TestSplineFitter.cpp`:

```cpp
    for (unsigned int i = 0; i < x.size(); i++)
        x[i] = i/(x.size()-1.0);
    double ya[] = {15.579, 16.235, 17.325, 18.741, 20.454, 22.517, 24.944, 27.554, 29.942, 31.657,
                   32.486, 32.612, 32.494, 32.532, 32.785, 32.917, 32.402, 30.842, 28.229, 24.989,
                   21.762, 19.074, 17.147, 15.970, 15.467, 15.579};
    y.assign(ya, ya+y.size());
    SplineFitter::createPeriodicSpline(x, y, deriv);
    ASSERT_EQUAL_TOL(SplineFitter::evaluateSplineDerivative(x, y, deriv, x[0]),
                     SplineFitter::evaluateSplineDerivative(x, y, deriv, x[x.size()-1]), 1e-6);
```

This raises the following error message:

```
exception: Assertion failure at TestSplineFitter.cpp:94.  Expected 9.88119, found 10.0784
```

**Comment by peastman** [comment]
> The method createNaturalSpline actually creates a clamped spline [1, 2], whose first-order derivative is set to be zero at the endpoints. For a natural spline, this should be done for the second-order derivative instead.

In this case, I think it comes out to the same thing.  TabulatedFunction defines its value to be 0 outside the range of the tabulated data.  If you think of the outside regions as just being extra segments to the curve, then the boundary conditions for a natural cubic spline would require both first and second derivatives to be 0 (in order to be continuous).

**Comment by craabreu** [comment]
Hi @peastman 

Unfortunately, there is not enough freedom in cubic spline to impose the values of both the first- and second-order derivatives at both boundaries. In fact, the first issue I pointed out is just a matter of terminology. One could, for instance, simply rename the method to `createClampedSpline`. IMHO, calling it `createNaturalSpline` causes a bit of confusion.

> TabulatedFunction defines its value to be 0 outside the range of the tabulated data.

Actually, `evaluateSpline` and `evaluateSplineDerivative` will throw an exception if we try to do any extrapolation. Do you think it is necessary to allow these methods to perform extrapolation? I'm asking this because I'm willing to implement any required changes.

> If you think of the outside regions as just being extra segments to the curve, then the boundary conditions for a natural cubic spline would require both first and second derivatives to be 0 (in order to be continuous).

In this case, at any point outside the range, the function should have the same value it has at the closest boundary, which is not necessarily zero, right? If this is the intention, a clamped spline is adequate, but it is worth noting that the second-order derivative will be discontinuous at the borders. Alternatively, we could actually use natural spline and extrapolate via straight lines whose inclinations are the same as determined at the boundaries. In this case, the function and its first and second derivatives would all be continuous everywhere. What do you think?

**Comment by peastman** [comment]
> One could, for instance, simply rename the method to createClampedSpline. IMHO, calling it createNaturalSpline causes a bit of confusion.

Good point.

> > TabulatedFunction defines its value to be 0 outside the range of the tabulated data.
>
> Actually, evaluateSpline and evaluateSplineDerivative will throw an exception if we try to do any extrapolation.

TabulatedFunction will never call it with a value outside the range:

https://github.com/openmm/openmm/blob/7c7dc7511446d05fe266c33389aea1db5d826a21/platforms/reference/src/ReferenceTabulatedFunction.cpp#L97-L102

> Alternatively, we could actually use natural spline and extrapolate via straight lines whose inclinations are the same as determined at the boundaries.

I think it's best to keep the current behavior outside the range.  For one thing, that would be a drastic change in behavior that could break a lot of code.  For another, having it be 0 outside the range is useful and easy to understand.  The linear extrapolation would produce values that the user can't even predict in advance, since they would depend sensitively on the (yet to be determined) derivative at the boundary.

**Comment by craabreu** [comment]
@peastman 
I'm sorry I misinterpreted the code of `createNaturalSpline`. It actually does what its name suggests, that is, natural spline. The problem is that the output argument `deriv` contains the second derivatives at the nodes, while I was assuming it contained the first ones. [This document](http://aero-comlab.stanford.edu/Papers/splines.pdf) contrasts these two formulations. I will now focus on discovering why `createPediocSpline` does not yield a continuous derivative.

**Comment by peastman** [comment]
This was fixed in #2629.
